src/rust/ide/enso-protocol/src/project_manager.rs

```diff
@@ -0,0 +1,382 @@
+            last_opened : Some(DateTime::parse_from_rfc3339("2020-02-02T13:15:20Z").unwrap())
+        };
+        let expected_recent_projects = ProjectListResponse { projects : vec![project1,project2] };
+        let sample1 = ProjectMetadata {
+            name        : ProjectName { name : "sample1".to_string() },
+            id          : Uuid::default(),
+            last_opened : Some(DateTime::parse_from_rfc3339("2019-11-23T05:30:12Z").unwrap())
+        };
+        let sample2 = ProjectMetadata {
+            name        : ProjectName { name : "sample2".to_string() },
+            id          : Uuid::default(),
+            last_opened : Some(DateTime::parse_from_rfc3339("2019-12-25T00:10:58Z").unwrap())
+        };
+        let expected_sample_projects = ProjectListResponse { projects : vec![sample1,sample2] };
+        mock_client.set_list_recent_projects_result(2,Ok(expected_recent_projects.clone()));
+        mock_client.set_list_samples_result(2,Ok(expected_sample_projects.clone()));
+
+        let list_recent_error = "Couldn't get recent projects.";
+        let list_sample_error = "Couldn't get sample projects.";
+        let recent_projects = result(mock_client.list_recent_projects(2)).expect(list_recent_error);
+        assert_eq!(recent_projects, expected_recent_projects);
+        let sample_projects = result(mock_client.list_samples(2)).expect(list_sample_error);
+        assert_eq!(sample_projects, expected_sample_projects);
+    }
+}
+
+
+
+// ====================
+// === Client tests ===
+// ====================
+
+#[cfg(test)]
+mod remote_client_tests {
+    use super::*;
+
+    use chrono::DateTime;
+    use json_rpc::messages::Message;
+    use json_rpc::messages::RequestMessage;
+    use json_rpc::test_util::transport::mock::MockTransport;
+    use serde_json::json;
+    use serde_json::Value;
+    use std::future::Future;
+    use utils::test::poll_future_output;
+    use futures::task::LocalSpawnExt;
+
+    struct Fixture {
+        transport : MockTransport,
+        client    : Client,
+        executor  : futures::executor::LocalPool,
+    }
+
+    fn setup_fm() -> Fixture {
+        let transport = MockTransport::new();
+        let client    = Client::new(transport.clone());
+        let executor  = futures::executor::LocalPool::new();
+        executor.spawner().spawn_local(client.runner()).unwrap();
+        Fixture {transport,client,executor}
+    }
+
+    /// Tests making a request using file manager:
+    /// * creates PM client and uses `make_request` to make a request
+    /// * checks that request is made for `expected_method`
+    /// * checks that request input is `expected_input`
+    /// * mocks receiving a response from server with `result`
+    /// * checks that FM-returned Future yields `expected_output`
+    fn test_request<Fun, Fut, T>
+    ( make_request:Fun
+      , expected_method:&str
+      , expected_input:&Value
+      , result:&Value
+      , expected_output:&T )
```

Comment:
alignment broken - `,` are indented too much + no alignment on `:`

Reply:
And closing ) in a new line.